spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,126 @@
+
+	private final ProtobufMessageConverter protobufMessageConverter;
+
+	private final Class<? extends com.google.protobuf.Message> defaultType;
+
+	private ClassLoader beanClassLoader;
+
+	private Expression typeIdExpression = new FunctionExpression<Message<?>>(
+			(message) -> message.getHeaders().get(ProtoHeaders.TYPE));
+
+	private EvaluationContext evaluationContext;
+
+	/**
+	 * Construct an instance with the supplied default type to create.
+	 * @param defaultType the type.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType) {
+		this(defaultType, new ProtobufMessageConverter());
+	}
+
+	/**
+	 * Construct an instance with the supplied default type and ProtobufMessageConverter instance.
+	 * @param defaultType the type.
+	 * @param protobufMessageConverter the message converter used.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType,
+			ProtobufMessageConverter protobufMessageConverter) {
+		Assert.notNull(defaultType, "'defaultType' must not be null");
+		this.defaultType = defaultType;
+		this.protobufMessageConverter = protobufMessageConverter;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpressionString(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	private void assertExpressionNotNull(Object expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+
+		Class<? extends com.google.protobuf.Message> targetClass = null;
+		Object value = this.typeIdExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			targetClass = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				targetClass = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (targetClass == null) {
+			targetClass = this.defaultType;
```

Comment:
I think we need at least `trace()` when we fallback to the default type.
Kinda say that `typeIdExpression` returned `null` for request message and therefore falling back to the default one.

Reply:
After thinking about this more I would suggest this behavior:
1. Change the `defaultType` from ctor option to the `expectedType` setter.
2. Have a ctor free from the type at all.
3. Still have an `expectedTypeExpression` and with that default value.
4. Have a setter for it anyway.
5. There is a `ValueExpression` for just constant `expectedType`.
6. If expression returns `null` we fail with an exception.

This way end-user by default can use this transformer against a header.
If other non-standard behavior is expected, you are free to override that expression or just set a static type.
My point is to avoid some unexpected logic, when we fallback to something what might not be relevant to the request at all.
You are free to provide a complex expression which would do a fallback for you, but that's already out of this transformer responsibility.
Another argument is to not require some default type when we are going to deal only with types resolved through expression.